# Changelog
All notable changes to **MetaSanity** and **BioMetaDB** will be documented in this file.

## MetaSanity v1.2.0 - 01/31/2020

### BioMetaDB v1.3.0
#### Changed
- BioMetaDB is now installable via `pip3`.
- Accessory scripts `table` and `tsv-join` part of `pip3` installation.

---
### MetaSanity-Code v0.0.6
#### Changed
- Refactor `install.py`, `MetaSanity.py` to use `pip3` version of BioMetaDB.

------
## MetaSanity v1.1.1 - 2020 version, 01/05/2020

### MetaSanity-Docker v0.1.2
- GTDB-Tk released a version update to fix bug in FastANI calling.

------
## MetaSanity v1.1.1 - 12/13/2019

### pipedm v0.0.5
#### Changed
- Updates to the user-provided genome folder (e.g. deleting or adding genome FASTA files) determines which genomes to annotate.
    - Will not take into account previous or failed runs that do not originate from this directory.
- Intermediary files are retained. These may be deleted by the user with the `-z` flag.
- CAZy outputs 'None' instead of '0', which is in line with rest of database API
    
#### Fixed
- Fixed bug in parsing genome names with multiple '.'
- Removed bug for genomes with no output

---
### BioMetaDB v0.1.2
#### Added
- Users can query for "unannotated" and "_unannot" from the command line
- Users can return truncated annotations (e.g., only return KEGG ID, etc) with `-r/--truncate`
    - Does not apply to prokka

#### Changed
- Checking for un/annotated status completed through simple existence check versus string comparison.

---
### MetaSanity-Docker v0.1.2
#### Changed
- CheckM and GTDB-Tk update to Python3.

#### Fixed
- InterProScan bug that triggers `lib.idn.so.1 missing`.
- Update Prokka `tbl2asn`.

------
## MetaSanity v1.1 - 10/18/2019

### pipedm v0.0.4
#### Added
- Incorporate tRNA/rRNA data in Prokka annotations.
- Determine high/medium/low quality genomes by incorporating Prokka annotations after running **PhyloSanity** and **FuncSanity**.

#### Changed
- Include gene description in annotations.
- Record names in `evaluation` table match table names in **BioMetaDB** project, less the file extension (`.fna` or `.protein.faa`).

---
### MetaSanity-Code v0.1.1
#### Changed
- Run Docker or SourceCode version from `MetaSanity.py`.
- Update `install.py`
    - Download Source Code or Docker version via this script.
    - Auto-install **MetaSanity** to download directory.
	    - Users no longer need to manually input their program installation location in the `MetaSanity.py` script.
- No more option to set name of program package or name of download directory - default to "MetaSanity" and "databases", respectively.
- All database table creation/updates are no longer handled in pipedm, but are now handled post-**MetaSanity** runs from within `MetaSanity.py`.

---
### BioMetaDB v0.1.1
#### Added
- Speed up `dbdm` runs for hundreds of genomes.
- Run table `UPDATE` functions from scripts and command line.
- Store hypothetical proteins.
- Filter out "hypothetical protein" annotations as part of "annotated" and "_annot" queries.

#### Changed
- `SUMMARIZE` info is not displayed to stdout for queries that write results to an external file or folder.

#### Fixed
- Correct type generation in update functions
- Correctly parses evaluation metrics for very low-quality genomes.
- Remove extra tab generated by the `SUMMARIZE` function when outputting annotations

------
## MetaSanity v1 - 10/01/2019
First Release


